snooty/parser.py

```diff
@@ -534,32 +436,120 @@ def handle_directive(
             try:
-                static_asset = self.add_static_asset(asset_path, False)
-                self.pending.append(PendingLiteralInclude(code, static_asset, options))
+                with open(filepath) as file:
+                    text = file.read()
+
             except OSError as err:
                 self.diagnostics.append(
-                    CannotOpenFile(argument_text, err.strerror, util.get_line(node))
+                    CannotOpenFile(argument_text, err.strerror, line)
+                )
+                return doc
+
+            lines = text.split("\n")
+
+            def _locate_text(text: str) -> int:
+                """
+                Searches the literally-included file ('lines') for the specified text. If no such text is found,
+                add an InvalidLiteralInclude diagnostic.
+                """
+                assert isinstance(text, str)
+                loc = next((idx for idx, line in enumerate(lines) if text in line), -1)
+                if loc < 0:
+                    self.diagnostics.append(
+                        InvalidLiteralInclude(f'"{text}" not found in {filepath}', line)
+                    )
+                return loc
+
+            # Locate the start_after query
+            start_after = 0
+            if "start-after" in options:
+                start_after_text = options["start-after"]
+                start_after = _locate_text(start_after_text)
+                # Only increment start_after if text is specified, to avoid capturing the start_after_text
+                start_after += 1
+
+            # ...now locate the end_before query
+            end_before = len(lines)
+            if "end-before" in options:
+                end_before_text = options["end-before"]
+                end_before = _locate_text(end_before_text)
+
+            # Check that start_after_text precedes end_before_text (and end_before exists)
+            if start_after >= end_before >= 0:
+                self.diagnostics.append(
+                    InvalidLiteralInclude(
+                        f'"{end_before_text}" precedes "{start_after_text}" in {filepath}',
+                        line,
+                    )
                 )
-            except ValueError as err:
-                msg = f'Invalid "literalinclude": {err}'
-                self.diagnostics.append(InvalidLiteralInclude(msg, util.get_line(node)))
-            return code
+
+            # If we failed to locate end_before text, default to the end-of-file
+            if end_before == -1:
+                end_before = len(lines)
+
+            # Capture the original file-length before splicing it
+            len_file = len(lines)
+            lines = lines[start_after:end_before]
+
+            dedent = 0
+            if "dedent" in options:
+                # Dedent is specified as a nonnegative integer (number of characters)
+                if type(options["dedent"]) is int:
```

Comment:
[nit] What about using `isinstance`?
```suggestion
                if isinstance(options["dedent"], int):
```

Reply:
This is intentional! Since `bool` is a subtype of `int`, and since `options["dedent"]` can take either type, there was an error in which this line was catching both types. I'll just swap the order instead, in order to preserve the `isinstance` call.